### PR TITLE
Add RSA Signature restrictions for X9.31 padding in the FIPS provider.

### PR DESCRIPTION
**Pull request by slontis**
In FIPS 140-3, RSA Signing with X931 padding is not approved,
but verification is allowed for legacy purposes. An indicator has been added
for RSA signing with X931 padding.

A strict restriction on the size of the RSA modulus has been added
i.e. It must be 1024 + 256 * s (which is part of the ANSI X9.31 spec).

Added implementation comments to the X931 padding code

##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by slontis** [comment]
Dependent on PR https://github.com/openssl/openssl/pull/24623
The first 5 commits should not be reviewed. This PR will be rebased once 24623 is merged.

**Review by paulidale — COMMENTED** [review]
Looks pretty good.
Will go through after the main PR is merged and this rebased.

File: providers/implementations/asymciphers/rsa_enc.c
Comment (t8m) [inline providers/implementations/asymciphers/rsa_enc.c:57]: Why ifndef FIPS_MODULE and instead just remove it? It is misleading and useless here, isn't it?

**Comment by slontis** [comment]
> Why ifndef FIPS_MODULE and instead just remove it? It is misleading and useless here, isn't it?

Should we completely remove X9.31 as a padding mode from RSA encryption (X9.31 was a signature standard)

**Comment by slontis** [comment]
CHANGES.md might need an update also for x9.31 padding with rsa encryption?

**Comment by slontis** [comment]
minor update for docs to use X9.31 rather than X931

**Comment by t8m** [comment]
> > Why ifndef FIPS_MODULE and instead just remove it? It is misleading and useless here, isn't it?
> 
> Should we completely remove X9.31 as a padding mode from RSA encryption (X9.31 was a signature standard)

IMO there is nothing for OTC to discuss here as RSA_X931_PADDING is not supported by rsa_ossl_private_decrypt() and rsa_ossl_public_encrypt() functions.

So you will just get the failure later when you call the encryption or decryption functions.

File: providers/implementations/asymciphers/rsa_enc.c
Comment (slontis) [reply providers/implementations/asymciphers/rsa_enc.c]: I have moved this to a separate PR as it does not belong here. That also removes the contention from this PR.

**Comment by slontis** [comment]
fixed some of the coding style issues.

**Review by paulidale — APPROVED** [review]

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
Provider compat CI failures are not relevant.

**Comment by slontis** [comment]
Rebased to fix merge conflicts - will merge if tests pass.

**Comment by slontis** [comment]
Re added files to fix it using the wrong parent. No changes performed otherwise.

**Review by paulidale — APPROVED** [review]

**Comment by slontis** [comment]
The errors are not related to this PR.

**Pull request closed without merging**
